archive/miu_language/arithmetic.lean

```diff
@@ -0,0 +1,47 @@
+/-
+Copyright (c) 2020 Gihan Marasingha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gihan Marasingha
+-/
+import data.nat.modeq
+import tactic.linarith
+
+/-!
+# Arithmetic
```

Comment:
I agree with your comment elsewhere that this file is now so short that it makes sense to make it a section in another file.
The lemmas don't seem directly useful outside of this project, but otherwise I would think they should actually move into the main `src/` tree of mathlib.

Reply:
OK. I'll do that. The one lemma that _might_ be more generally useful is the fixed-point result:
```lean
lemma eq_of_countI_eq_length  {ys : miustr} (h : count I ys = length ys) :  ys = repeat I (count I ys) :=
```

Of course, this should be expressed in its generalisation:
```lean
variable α : Type*
variable [decidable_eq α]

lemma eq_of_count_eq_length {a : α} {l : list α} (h : count a l = length l) :  l = repeat a (count a l) :=
```

But I imagine that including every slightly useful lemma would cause mathlib to grow exponentially!

Reply:
Other potentially useful lemmas (when suitably generalised) are:

```lean
lemma tail_append_singleton_of_ne_nil {a : miu_atom} {as : miustr} (h₁ : as ≠ nil) : 
  tail (as ++ [a]) = tail as ++ [a] :=
```
and
```lean
lemma exists_cons_of_ne_nil {as : miustr} (h : as ≠ nil) : ∃ c cs, as = c :: cs :=
```

Reply:
I think all three of these are fine for the main `src/` directory!